modules/ngx_http_lua_upstream/src/ngx_http_lua_upstream_module.c

```diff
@@ -0,0 +1,654 @@
+    lua_pushinteger(L, (lua_Integer) peer->max_fails);
+    lua_rawset(L, -3);
+
+    lua_pushliteral(L, "fail_timeout");
+    lua_pushinteger(L, (lua_Integer) peer->fail_timeout);
+    lua_rawset(L, -3);
+
+    if (peer->accessed) {
+        lua_pushliteral(L, "accessed");
+        lua_pushinteger(L, (lua_Integer) peer->accessed);
+        lua_rawset(L, -3);
+    }
+
+    if (peer->checked) {
+        lua_pushliteral(L, "checked");
+        lua_pushinteger(L, (lua_Integer) peer->checked);
+        lua_rawset(L, -3);
+    }
+
+    if (peer->down) {
+        lua_pushliteral(L, "down");
+        lua_pushboolean(L, 1);
+        lua_rawset(L, -3);
+    }
+
+    return 0;
+}
+
+
+static ngx_http_upstream_main_conf_t *
+ngx_http_lua_upstream_get_upstream_main_conf(lua_State *L)
+{
+    ngx_http_request_t                   *r;
+
+    r = ngx_http_lua_get_request(L);
+
+    if (r == NULL) {
+        return ngx_http_cycle_get_module_main_conf(ngx_cycle,
+                                                   ngx_http_upstream_module);
+    }
+
+    return ngx_http_get_module_main_conf(r, ngx_http_upstream_module);
+}
+
+
+static ngx_http_upstream_srv_conf_t *
+ngx_http_lua_upstream_find_upstream(lua_State *L, ngx_str_t *host)
+{
+    u_char                               *port;
+    size_t                                len;
+    ngx_int_t                             n;
+    ngx_uint_t                            i;
+    ngx_http_upstream_srv_conf_t        **uscfp, *uscf;
+    ngx_http_upstream_main_conf_t        *umcf;
+
+    umcf = ngx_http_lua_upstream_get_upstream_main_conf(L);
+
+#if (NGX_HTTP_UPSTREAM_RBTREE)
+
+    ngx_list_part_t                *part;
+
+    uscf = ngx_http_upstream_rbtree_lookup(umcf, host);
+
+    if (uscf != NULL)
+    {
+        return uscf;
+    }
+
+    part = &umcf->implicit_upstreams.part;
+    uscfp = part->elts;
+
+    for (i = 0; /* void */ ; i++) {
+
+        if (i >= part->nelts) {
+            if (part->next == NULL) {
+                break;
+            }
+
+            part = part->next;
+            uscfp = part->elts;
+            i = 0;
+        }
+
+#else
+
+    uscfp = umcf->upstreams.elts;
+
+    for (i = 0; i < umcf->upstreams.nelts; i++) {
+
+#endif
+
+        uscf = uscfp[i];
+
+        if (uscf->host.len == host->len
+            && ngx_memcmp(uscf->host.data, host->data, host->len) == 0)
+        {
+            return uscf;
+        }
+    }
+
+    port = ngx_strlchr(host->data, host->data + host->len, ':');
+    if (port) {
+        port++;
+        n = ngx_atoi(port, host->data + host->len - port);
+        if (n < 1 || n > 65535) {
+            return NULL;
+        }
+
+        /* try harder with port */
+
+        len = port - host->data - 1;
+
+#if (NGX_HTTP_UPSTREAM_RBTREE)
+        ngx_str_t addr;
+        addr.data = host->data;
+        addr.len = len;
+        uscf = ngx_http_upstream_rbtree_lookup(umcf, &addr);
+        if (uscf != NULL && uscf->port
+            && uscf->port == n)
+        {
```

Comment:
can be included in one line, it does not exceed 79 chars

Reply:
@chobits fixed